### PR TITLE
Add sharedRef support for async function returns to expo-modules-core

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

We want to implement an async `load` method that returns a sharedRef on promise complete.

# How

Add sharedRefs as one of the invokable types.
Change type of CallbackArg to support either folly:dynamic or sharedRef arguments.
Add a class registry to store native class to js object mapping.
Use that class registry to create new JS objects when returning a sharedRef from an async fn.

# Test Plan

Tested in bare-expo on Android.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by lukmccall — COMMENTED** [review]
I left some suggestions after my first reading ;) It's looking good, some rough edges that need polish.

File: packages/expo-modules-core/android/src/main/cpp/JavaCallback.h [hunk 1]
```diff
@@ -6,23 +6,55 @@
...
 namespace react = facebook::react;
 
 namespace expo {
+
+struct SharedObjectId : public jni::HybridClass<SharedObjectId> {
+  static constexpr const char *kJavaDescriptor = "Lexpo/modules/kotlin/sharedobjects/SharedObjectId;";
+
+  SharedObjectId();
+
+  RN_EXPORT int value{};
```
Comment (lukmccall) [inline packages/expo-modules-core/android/src/main/cpp/JavaCallback.h:25]:
```suggestion
  int value{};
```
Is that needed? If so, what's this doing?

**Review by lukmccall — APPROVED** [review]

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/classcomponent/ClassRegistry.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package expo.modules.kotlin.sharedobjects
+
+import expo.modules.kotlin.AppContext
+import expo.modules.kotlin.jni.JavaScriptObject
+
+
+
```
Comment (tsapeta) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/classcomponent/ClassRegistry.kt:7]:
```suggestion
```

**Review by tsapeta — APPROVED** [review]

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/classcomponent/ClassRegistry.kt (hunk 2 above)
Comment (aleqsio) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/classcomponent/ClassRegistry.kt:7]:
```suggestion
```

**Pull request merged**